File: kyc-kyb-system/liveness-service/app/middleware/rbac.py
"""
RBAC Middleware for Liveness Detection Service
Integrates Keycloak authentication and Permify authorization
"""

import os
import logging
from typing import Optional, List
from functools import wraps

import jwt
import requests
from fastapi import Request, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

logger = logging.getLogger(__name__)

# Configuration
KEYCLOAK_URL = os.getenv("KEYCLOAK_URL", "http://localhost:8080")
KEYCLOAK_REALM = os.getenv("KEYCLOAK_REALM", "kyc-kyb-system")
PERMIFY_URL = os.getenv("PERMIFY_URL", "http://localhost:3476")

# Keycloak public key cache
_keycloak_public_key = None

security = HTTPBearer()


class RBACMiddleware:
    """RBAC Middleware for authentication and authorization"""
    
    def __init__(self):
        self.keycloak_url = KEYCLOAK_URL
        self.realm = KEYCLOAK_REALM
        self.permify_url = PERMIFY_URL
        self.public_key = None
        
    def get_keycloak_public_key(self) -> str:
        """Get Keycloak public key for JWT verification"""
        global _keycloak_public_key
        
        if _keycloak_public_key:
            return _keycloak_public_key
            
        try:
            url = f"{self.keycloak_url}/realms/{self.realm}"
            response = requests.get(url, timeout=5)
            response.raise_for_status()
            
            realm_info = response.json()
            public_key = realm_info.get("public_key")
            
            if not public_key:
                raise ValueError("Public key not found in realm info")
                
            _keycloak_public_key = f"-----BEGIN PUBLIC KEY-----\n{public_key}\n-----END PUBLIC KEY-----"
            return _keycloak_public_key
            
        except Exception as e:
            logger.error(f"Failed to get Keycloak public key: {e}")
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail="Authentication service unavailable"
            )
    
    def verify_token(self, token: str) -> dict:
        """Verify JWT token from Keycloak"""
        try:
            public_key = self.get_keycloak_public_key()
            
            payload = jwt.decode(
                token,
                public_key,
                algorithms=["RS256"],
                audience="account",
                options={"verify_exp": True}
            )
            
            return payload
            
        except jwt.ExpiredSignatureError:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token has expired"
            )
        except jwt.InvalidTokenError as e:
            logger.error(f"Invalid token: {e}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid authentication token"
            )
    
    def extract_user_info(self, payload: dict) -> dict:
        """Extract user information from JWT payload"""
        return {
            "user_id": payload.get("sub"),
            "username": payload.get("preferred_username"),
            "email": payload.get("email"),
            "roles": payload.get("realm_access", {}).get("roles", []),
            "organization_id": payload.get("organization_id"),
            "name": payload.get("name"),
        }
    
    def check_permission(
        self,
        user_id: str,
        permission: str,
        resource_id: Optional[str] = None
    ) -> bool:
        """Check permission using Permify"""
        try:
            url = f"{self.permify_url}/v1/permissions/check"
            
            # Determine entity type based on permission
            if permission.startswith("liveness"):
                entity_type = "liveness_check"
            else:
                entity_type = "user"
            
            payload = {
                "entity": {
                    "type": entity_type,
                    "id": resource_id or "default"
                },
                "permission": permission,
                "subject": {
                    "type": "user",
                    "id": user_id
                }
            }
            
            response = requests.post(url, json=payload, timeout=2)
            
            if response.status_code == 200:
                result = response.json()
                return result.get("can", False)
            else:
                logger.warning(f"Permify check failed: {response.status_code}")
                return False
                
        except Exception as e:
            logger.error(f"Permission check failed: {e}")
            # Fail open for now (should be fail closed in production)
            return False
    
    def check_role_permission(self, roles: List[str], required_roles: List[str]) -> bool:
        """Check if user has any of the required roles"""
        return any(role in roles for role in required_roles)
    
    async def authenticate(self, request: Request) -> dict:
        """Authenticate request and extract user info"""
        auth_header = request.headers.get("Authorization")
        
        if not auth_header or not auth_header.startswith("Bearer "):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Missing or invalid authorization header"
            )
        
        token = auth_header.split(" ")[1]
        payload = self.verify_token(token)
        user_info = self.extract_user_info(payload)
        
        # Attach user info to request state
        request.state.user = user_info
        
        return user_info


# Global middleware instance
rbac_middleware = RBACMiddleware()


def require_auth(func):
    """Decorator to require authentication"""
    @wraps(func)
    async def wrapper(*args, **kwargs):
        # Extract request from args
        request = None
        for arg in args:
            if isinstance(arg, Request):
                request = arg
                break
        
        if not request:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Request object not found"
            )
        
        # Authenticate
        user_info = await rbac_middleware.authenticate(request)
        
        # Call original function
        return await func(*args, **kwargs)
    
    return wrapper


def require_roles(required_roles: List[str]):
    """Decorator to require specific roles"""
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            # Extract request from args
            request = None
            for arg in args:
                if isinstance(arg, Request):
                    request = arg
                    break
            
            if not request:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail="Request object not found"
                )
            
            # Authenticate
            user_info = await rbac_middleware.authenticate(request)
            
            # Check roles
            user_roles = user_info.get("roles", [])
            
            # System administrator has access to everything
            if "system_administrator" in user_roles:
                return await func(*args, **kwargs)
            
            # Check required roles
            if not rbac_middleware.check_role_permission(user_roles, required_roles):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail=f"Insufficient permissions. Required roles: {required_roles}"
                )
            
            return await func(*args, **kwargs)
        
        return wrapper
    return decorator


def require_permission(permission: str):
    """Decorator to require specific permission via Permify"""
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            # Extract request from args
            request = None
            for arg in args:
                if isinstance(arg, Request):
                    request = arg
                    break
            
            if not request:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail="Request object not found"
                )
            
            # Authenticate
            user_info = await rbac_middleware.authenticate(request)
            user_id = user_info.get("user_id")
            
            # System administrator has access to everything
            if "system_administrator" in user_info.get("roles", []):
                return await func(*args, **kwargs)
            
            # Check permission via Permify
            resource_id = kwargs.get("id") or kwargs.get("customer_id")
            
            if not rbac_middleware.check_permission(user_id, permission, resource_id):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail=f"Insufficient permissions. Required: {permission}"
                )
            
            return await func(*args, **kwargs)
        
        return wrapper
    return decorator


# Permission constants for Liveness Service
class LivenessPermissions:
    CHECK_CREATE = "liveness.check.create"
    CHECK_READ = "liveness.check.read"
    FACE_MATCH = "liveness.face.match"
    FEATURES_EXTRACT = "liveness.features.extract"


# Role constants
class Roles:
    SYSTEM_ADMIN = "system_administrator"
    COMPLIANCE_OFFICER = "compliance_officer"
    KYC_ANALYST = "kyc_analyst"
    RISK_MANAGER = "risk_manager"
    KYC_OPERATOR = "kyc_operator"

File: kyc-kyb-system/risk-scoring-service/internal/middleware/rbac.go
package middleware

import (
	"context"
	"crypto/rsa"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
)

var (
	keycloakURL   = getEnv("KEYCLOAK_URL", "http://localhost:8080")
	keycloakRealm = getEnv("KEYCLOAK_REALM", "kyc-kyb-system")
	permifyURL    = getEnv("PERMIFY_URL", "http://localhost:3476")
	publicKey     *rsa.PublicKey
	publicKeyOnce sync.Once
)

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

// UserInfo contains authenticated user information
type UserInfo struct {
	UserID         string   `json:"user_id"`
	Username       string   `json:"username"`
	Email          string   `json:"email"`
	Roles          []string `json:"roles"`
	OrganizationID string   `json:"organization_id"`
	Name           string   `json:"name"`
}

// Claims represents JWT claims
type Claims struct {
	Sub               string   `json:"sub"`
	PreferredUsername string   `json:"preferred_username"`
	Email             string   `json:"email"`
	Name              string   `json:"name"`
	OrganizationID    string   `json:"organization_id"`
	RealmAccess       struct {
		Roles []string `json:"roles"`
	} `json:"realm_access"`
	jwt.RegisteredClaims
}

// getPublicKey retrieves Keycloak public key
func getPublicKey() (*rsa.PublicKey, error) {
	var err error
	publicKeyOnce.Do(func() {
		url := fmt.Sprintf("%s/realms/%s", keycloakURL, keycloakRealm)
		
		resp, httpErr := http.Get(url)
		if httpErr != nil {
			err = fmt.Errorf("failed to get realm info: %w", httpErr)
			return
		}
		defer resp.Body.Close()

		body, httpErr := io.ReadAll(resp.Body)
		if httpErr != nil {
			err = fmt.Errorf("failed to read response: %w", httpErr)
			return
		}

		var realmInfo map[string]interface{}
		if httpErr := json.Unmarshal(body, &realmInfo); httpErr != nil {
			err = fmt.Errorf("failed to parse realm info: %w", httpErr)
			return
		}

		publicKeyStr, ok := realmInfo["public_key"].(string)
		if !ok {
			err = fmt.Errorf("public key not found in realm info")
			return
		}

		pemKey := fmt.Sprintf("-----BEGIN PUBLIC KEY-----\n%s\n-----END PUBLIC KEY-----", publicKeyStr)
		publicKey, err = jwt.ParseRSAPublicKeyFromPEM([]byte(pemKey))
	})

	return publicKey, err
}

// verifyToken verifies JWT token
func verifyToken(tokenString string) (*Claims, error) {
	pubKey, err := getPublicKey()
	if err != nil {
		return nil, err
	}

	token, err := jwt.ParseWithClaims(tokenString, &Claims{}, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return pubKey, nil
	})

	if err != nil {
		return nil, err
	}

	if claims, ok := token.Claims.(*Claims); ok && token.Valid {
		return claims, nil
	}

	return nil, fmt.Errorf("invalid token")
}

// extractUserInfo extracts user information from claims
func extractUserInfo(claims *Claims) *UserInfo {
	return &UserInfo{
		UserID:         claims.Sub,
		Username:       claims.PreferredUsername,
		Email:          claims.Email,
		Roles:          claims.RealmAccess.Roles,
		OrganizationID: claims.OrganizationID,
		Name:           claims.Name,
	}
}

// checkPermission checks permission using Permify
func checkPermission(userID, permission, resourceID string) bool {
	if resourceID == "" {
		resourceID = "default"
	}

	entityType := "risk_score"
	if strings.HasPrefix(permission, "risk") {
		entityType = "risk_score"
	}

	payload := map[string]interface{}{
		"entity": map[string]string{
			"type": entityType,
			"id":   resourceID,
		},
		"permission": permission,
		"subject": map[string]string{
			"type": "user",
			"id":   userID,
		},
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		log.Printf("Failed to marshal permission check payload: %v", err)
		return false
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "POST", permifyURL+"/v1/permissions/check", strings.NewReader(string(jsonData)))
	if err != nil {
		log.Printf("Failed to create permission check request: %v", err)
		return false
	}

	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Printf("Permission check request failed: %v", err)
		return false
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		log.Printf("Permission check failed with status: %d", resp.StatusCode)
		return false
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		log.Printf("Failed to decode permission check response: %v", err)
		return false
	}

	can, ok := result["can"].(bool)
	return ok && can
}

// hasRole checks if user has any of the required roles
func hasRole(userRoles []string, requiredRoles []string) bool {
	for _, required := range requiredRoles {
		for _, userRole := range userRoles {
			if userRole == required {
				return true
			}
		}
	}
	return false
}

// AuthMiddleware authenticates requests
func AuthMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" || !strings.HasPrefix(authHeader, "Bearer ") {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Missing or invalid authorization header"})
			c.Abort()
			return
		}

		tokenString := strings.TrimPrefix(authHeader, "Bearer ")

		claims, err := verifyToken(tokenString)
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid authentication token"})
			c.Abort()
			return
		}

		userInfo := extractUserInfo(claims)
		c.Set("user", userInfo)
		c.Next()
	}
}

// RequireRoles middleware requires specific roles
func RequireRoles(requiredRoles ...string) gin.HandlerFunc {
	return func(c *gin.Context) {
		userInfoInterface, exists := c.Get("user")
		if !exists {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
			c.Abort()
			return
		}

		userInfo, ok := userInfoInterface.(*UserInfo)
		if !ok {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Invalid user info"})
			c.Abort()
			return
		}

		// System administrator has access to everything
		if hasRole(userInfo.Roles, []string{"system_administrator"}) {
			c.Next()
			return
		}

		// Check required roles
		if !hasRole(userInfo.Roles, requiredRoles) {
			c.JSON(http.StatusForbidden, gin.H{
				"error": fmt.Sprintf("Insufficient permissions. Required roles: %v", requiredRoles),
			})
			c.Abort()
			return
		}

		c.Next()
	}
}

// RequirePermission middleware requires specific permission via Permify
func RequirePermission(permission string) gin.HandlerFunc {
	return func(c *gin.Context) {
		userInfoInterface, exists := c.Get("user")
		if !exists {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
			c.Abort()
			return
		}

		userInfo, ok := userInfoInterface.(*UserInfo)
		if !ok {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Invalid user info"})
			c.Abort()
			return
		}

		// System administrator has access to everything
		if hasRole(userInfo.Roles, []string{"system_administrator"}) {
			c.Next()
			return
		}

		// Get resource ID from URL params
		resourceID := c.Param("id")
		if resourceID == "" {
			resourceID = c.Param("customer_id")
		}

		// Check permission via Permify
		if !checkPermission(userInfo.UserID, permission, resourceID) {
			c.JSON(http.StatusForbidden, gin.H{
				"error": fmt.Sprintf("Insufficient permissions. Required: %s", permission),
			})
			c.Abort()
			return
		}

		c.Next()
	}
}

// GetUserInfo retrieves user info from context
func GetUserInfo(c *gin.Context) (*UserInfo, error) {
	userInfoInterface, exists := c.Get("user")
	if !exists {
		return nil, fmt.Errorf("user not authenticated")
	}

	userInfo, ok := userInfoInterface.(*UserInfo)
	if !ok {
		return nil, fmt.Errorf("invalid user info")
	}

	return userInfo, nil
}

// Permission constants for Risk Service
const (
	PermissionRiskScoreCreate   = "risk.score.create"
	PermissionRiskScoreRead     = "risk.score.read"
	PermissionRiskScoreOverride = "risk.score.override"
)

// Role constants
const (
	RoleSystemAdmin       = "system_administrator"
	RoleComplianceOfficer = "compliance_officer"
	RoleKYCAnalyst        = "kyc_analyst"
	RoleRiskManager       = "risk_manager"
	RoleKYCOperator       = "kyc_operator"
)
